zerocopy-derive/src/lib.rs

```diff
@@ -587,7 +596,33 @@ fn impl_block<D: DataExt>(
         GenericParam::Const(cnst) => quote!(#cnst),
     });
 
-    let is_bit_valid = emit_is_bit_valid.then(|| {
+    let is_bit_valid = is_bit_valid_opts.map(|opts| {
+        let validator = Option::from(opts)
+            .map(|ZcValidatorAttr(validator)| {
+                quote::quote_spanned!(validator.span()=> && {
+                    // This assignment helps make the compiler error more precise if
+                    // `#validator` doesn't return a `bool`. We use `quote_spanned!`
+                    // so that the error is attributed to the annotation rather than
+                    // just to the `TryFromBytes` token in
+                    // `#[derive(TryFromBytes)]`:
+                    //
+                    //   #[zerocopy(validator = "validate")]
+                    //                          ^^^^^^^^^^
+                    //                               |
+                    //                               |
+                    //   (type errors attributed to this span)
+                    //
+                    // TODO: Validate that assigning to this function pointer has no
+                    // performance impact. Presumably it's trivially
+                    // constant-foldable, but since it's a function pointer, maybe
+                    // the compiler is worse at optimizing it?
+                    let validate: fn(&Self) -> bool = Self::#validator;
```

Comment:
Support functions that are not methods; the user can themselves write `Self::validator` if need be.

Reply:
Better yet, support arbitrary expressions:

```rust
#[zerocopy(validator = |x| x.foo())]
```

Emit code in a wrapped function to prevent validator from messing with environment:

```rust
fn is_bit_valid(candidate: NonNull<Self>) -> bool {
    let c: &Self = {
        // unsafe validation stuff
    };

    // Even if hygiene is broken, this can't access anything from inside
    // the preceding block.
    let validate: fn(&Self) -> bool = #validator;
    validate(c)
}
```